### PR TITLE
optimize memory

### PR DESCRIPTION
**Pull request by 516108736**
1.delete Header()
2.modify  getCoinbaseAddressUntilBlock

### CODE REVIEW

**Comment by 516108736** [comment]
修改后的checkdb用时12小时1分钟 ：https://github.com/516108736/goquarkchain/runs/512911810?check_suite_focus=true
没改之前用时11小时14分支：https://github.com/QuarkChain/goquarkchain/runs/507098300?check_suite_focus=true

File: cluster/slave/api_backend.go [hunk 1]
```diff
@@ -425,14 +425,14 @@ func (s *SlaveBackend) getMinorBlockHeadersWithSkip(gReq *p2p.GetMinorBlockHeade
 		mTip   = shrd.MinorBlockChain.CurrentBlock()
 	)
 	if gReq.Type == qcom.SkipHash {
-		iHeader := shrd.MinorBlockChain.GetHeader(gReq.GetHash())
+		iHeader := shrd.MinorBlockChain.GetBlock(gReq.GetHash())
```
Comment (pingke) [inline cluster/slave/api_backend.go]: iBlock := shrd.MinorBlockChain.GetBlock(gReq.GetHash())
		if qcom.IsNil(iBlock) {
			return nil, fmt.Errorf("failed to get minor block header by hash, minor hash: %s", gReq.GetHash().String())
		}
		height = iBlock.NumberU64()
		hash, err := shrd.MinorBlockChain.GetHashByHeight(&height)
		if err != nil || hash != iBlock.Hash() {
			return nil, fmt.Errorf("failed to get minor block header by number or hash dont match, minor number: %d", height)
		}

File: consensus/posw/posw.go [hunk 2]
```diff
@@ -143,19 +155,19 @@ func (p *PoSW) getCoinbaseAddressUntilBlock(headerHash common.Hash) ([]account.R
 	return addrs, nil
 }
 
-func (p *PoSW) GetPoSWInfo(header types.IHeader, stakes *big.Int, address account.Recipient) (effectiveDiff *big.Int, mineable, mined uint64, err error) {
+func (p *PoSW) GetPoSWInfo(header types.IBlock, stakes *big.Int, address account.Recipient) (effectiveDiff *big.Int, mineable, mined uint64, err error) {
```
Comment (pingke) [inline consensus/posw/posw.go]: rename header to block

File: core/xshard_tx_cursor.go [hunk 3]
```diff
@@ -15,15 +15,15 @@ type blockchain interface {
 	GetBranch() account.Branch
 	GetGenesisRootHeight() uint32
 	GetRootBlockByHash(hash common.Hash) *types.RootBlock
-	GetRootBlockHeaderByHeight(h common.Hash, height uint64) *types.RootBlockHeader
+	GetRootBlockHeaderByHeight(h common.Hash, height uint64) *types.RootBlock
```
Comment (pingke) [inline core/xshard_tx_cursor.go]: rename

File: core/xshard_tx_cursor.go [hunk 4]
```diff
@@ -15,15 +15,15 @@ type blockchain interface {
...
 	ReadCrossShardTxList(hash common.Hash) *types.CrossShardTransactionDepositList
 	isNeighbor(remoteBranch account.Branch, rootHeight *uint32) bool
 }
 
 type XShardTxCursor struct {
 	bc                 blockchain
-	mBlockHeader       *types.MinorBlockHeader
-	maxRootBlockHeader *types.RootBlockHeader
+	mBlockHeader       *types.MinorBlock
```
Comment (pingke) [inline core/xshard_tx_cursor.go]: rename

File: core/xshard_tx_cursor.go [hunk 5]
```diff
@@ -40,14 +40,14 @@ type XShardTxCursor struct {
    # - EOF
    # - A valid x-shard transaction deposit
 */
-func NewXShardTxCursor(bc blockchain, mBlockHeader *types.MinorBlockHeader, cursorInfo *types.XShardTxCursorInfo) *XShardTxCursor {
+func NewXShardTxCursor(bc blockchain, mBlockHeader *types.MinorBlock, cursorInfo *types.XShardTxCursorInfo) *XShardTxCursor {
```
Comment (pingke) [inline core/xshard_tx_cursor.go]: rename

File: core/minorblockchain_addon.go [hunk 6]
```diff
@@ -114,12 +114,12 @@ func (m *MinorBlockChain) putMinorBlock(mBlock *types.MinorBlock, xShardReceiveT
 }
 
 func (m *MinorBlockChain) updateTip(state *state.StateDB, block *types.MinorBlock) (bool, error) {
-	preRootHeader := m.getRootBlockHeaderByHash(block.PrevRootBlockHash())
+	preRootHeader := m.GetRootBlockByHash(block.PrevRootBlockHash())
```
Comment (pingke) [inline core/minorblockchain_addon.go]: rename

File: consensus/posw/posw.go [hunk 7]
```diff
@@ -3,18 +3,19 @@ package posw
...
 	"github.com/QuarkChain/goquarkchain/core/types"
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/log"
 	lru "github.com/hashicorp/golang-lru"
-	"math/big"
 )
 
 type headReader interface {
-	GetHeader(hash common.Hash) types.IHeader
+	GetBlock(hash common.Hash) types.IBlock
```
[the remark this answers is not in the document]
Comment (516108736) [reply consensus/posw/posw.go]: offline discuss , not change back now

File: consensus/posw/posw.go [hunk 8]
```diff
@@ -88,8 +88,8 @@ func (p *PoSW) BuildSenderDisallowMap(headerHash common.Hash, coinbase *account.
 	return disallowMap, nil
 }
 
-func (p *PoSW) IsPoSWEnabled(header types.IHeader) bool {
-	return p.config.Enabled && header.GetTime() >= p.config.EnableTimestamp && header.NumberU64() > 0
+func (p *PoSW) IsPoSWEnabled(time uint64, height uint64) bool {
```
[the remark this answers is not in the document]
Comment (516108736) [reply consensus/posw/posw.go]: 这个地方已经改回去了，但是建议不要改回去，可以再看下

File: core/minorblock_validator.go [hunk 9]
```diff
@@ -99,13 +99,13 @@ func (v *MinorBlockValidator) ValidateBlock(mBlock types.IBlock, force bool) err
 		return ErrPrunedAncestor
 	}
 
-	prevHeader := v.bc.GetHeader(block.ParentHash())
-	if common.IsNil(prevHeader) {
+	prevBlock := v.bc.GetBlock(block.ParentHash())
```
Comment (pingke) [inline core/minorblock_validator.go]: v.bc.GetMinorBlock(block.ParentHash()) ?

File: core/minorblock_validator.go [hunk 10]
```diff
@@ -171,39 +171,39 @@ func (v *MinorBlockValidator) ValidateBlock(mBlock types.IBlock, force bool) err
...
 		return errMustBeOneMinorChain
 	}
 
-	if !v.bc.isSameRootChain(v.bc.getRootBlockHeaderByHash(block.PrevRootBlockHash()),
-		v.bc.getRootBlockHeaderByHash(prevHeader.(*types.MinorBlockHeader).GetPrevRootBlockHash())) {
+	if !v.bc.isSameRootChain(v.bc.GetRootBlockByHash(block.PrevRootBlockHash()),
+		v.bc.GetRootBlockByHash(prevBlock.IHeader().(*types.MinorBlockHeader).GetPrevRootBlockHash())) {
 		errMustBeOneRootChain := errors.New("prev root blocks are not on the same chain")
-		log.Error(v.logInfo, "err", errMustBeOneRootChain, "long", block.PrevRootBlockHash().String(), "short", prevHeader.(*types.MinorBlockHeader).GetPrevRootBlockHash().String())
+		log.Error(v.logInfo, "err", errMustBeOneRootChain, "long", block.PrevRootBlockHash().String(), "short", prevBlock.IHeader().(*types.MinorBlockHeader).GetPrevRootBlockHash().String())
```
Comment (pingke) [inline core/minorblock_validator.go]: prevBlock.GetPrevRootBlockHash().String()

File: consensus/posw/posw.go [hunk 11]
```diff
@@ -143,19 +155,19 @@ func (p *PoSW) getCoinbaseAddressUntilBlock(headerHash common.Hash) ([]account.R
...
-	blockCnt, err := p.countCoinbaseBlockUntil(header.Hash(), address)
+func (p *PoSW) GetPoSWInfo(block types.IBlock, stakes *big.Int, address account.Recipient) (effectiveDiff *big.Int, mineable, mined uint64, err error) {
+	blockCnt, err := p.countCoinbaseBlockUntil(block.Hash(), address)
 	if err != nil {
-		return header.GetDifficulty(), 0, 0, err
+		return block.Difficulty(), 0, 0, err
 	}
-	if !p.IsPoSWEnabled(header) || stakes == nil {
-		return header.GetDifficulty(), 0, blockCnt, nil
+	if !p.IsPoSWEnabled(block.IHeader()) || stakes == nil {
```
Comment (pingke) [inline consensus/posw/posw.go]: change IsPoSWEnabled also?

File: core/minorblockchain_addon.go [hunk 12]
```diff
@@ -1718,16 +1730,16 @@ func (m *MinorBlockChain) PoswInfo(mBlock *types.MinorBlock) (*rpc.PoSWInfo, err
 	if mBlock == nil {
 		return nil, errors.New("get powInfo err:mBlock is full")
 	}
-	header := mBlock.Header()
-	if !m.posw.IsPoSWEnabled(header) {
+
+	if !m.posw.IsPoSWEnabled(mBlock.Header()) {
```
Comment (pingke) [inline core/minorblockchain_addon.go]: update IsPoSWEnabled

File: core/rootblockchain.go [hunk 13]
```diff
@@ -1423,15 +1415,15 @@ func (bc *RootBlockChain) GetRootChainStakesFunc() func(address account.Address,
 	return bc.rootChainStakesFunc
 }
 
-func (bc *RootBlockChain) PoSWInfo(header *types.RootBlockHeader) (*rpc.PoSWInfo, error) {
-	if header.Number == 0 {
+func (bc *RootBlockChain) PoSWInfo(block *types.RootBlock) (*rpc.PoSWInfo, error) {
```
[the remark this answers is not in the document]
Comment (516108736) [reply core/rootblockchain.go]: changfe back

File: core/minorblockchain_addon.go (hunk 12 above)
Comment (516108736) [reply core/minorblockchain_addon.go]: offline discuss
use	IsPoSWEnabled(time uint64, height uint64) bool

File: core/minorblockchain_addon.go [hunk 14]
```diff
@@ -463,29 +464,29 @@ func (m *MinorBlockChain) reRunBlockWithState(block *types.MinorBlock) error {
 }
 
 // getEvmStateForNewBlock get evmState for new block.should have locked
-func (m *MinorBlockChain) getEvmStateForNewBlock(mHeader types.IHeader, ephemeral bool) (*state.StateDB, error) {
-	prevHash := mHeader.GetParentHash()
+func (m *MinorBlockChain) getEvmStateForNewBlock(mHeader *types.MinorBlock, ephemeral bool) (*state.StateDB, error) {
```
Comment (pingke) [inline core/minorblockchain_addon.go]: rename mHeader

File: core/minorblockchain_addon.go [hunk 15]
```diff
@@ -463,29 +464,29 @@ func (m *MinorBlockChain) reRunBlockWithState(block *types.MinorBlock) error {
...
 		evmState = evmState.Copy()
 	}
-	m.setEvmStateWithHeader(evmState, mHeader.(*types.MinorBlockHeader))
+	m.setEvmStateWithHeader(evmState, mHeader)
 	return evmState, nil
 }
 
-func (m *MinorBlockChain) setEvmStateWithHeader(evmState *state.StateDB, header *types.MinorBlockHeader) {
-	evmState.SetTimeStamp(header.GetTime())
+func (m *MinorBlockChain) setEvmStateWithHeader(evmState *state.StateDB, header *types.MinorBlock) {
```
Comment (pingke) [inline core/minorblockchain_addon.go]: rename header

File: core/minorblockchain_addon.go [hunk 16]
```diff
@@ -923,7 +924,8 @@ func (m *MinorBlockChain) AddRootBlock(rBlock *types.RootBlock) (bool, error) {
 		prevRootHeader := m.GetRootBlockByHash(mHeader.PrevRootBlockHash)
```
Comment (pingke) [inline core/minorblockchain_addon.go]: rename prevRootHeader

File: core/minorblockchain_addon.go (hunk 12 above)
Comment (pingke) [inline core/minorblockchain_addon.go]: update  IsPoSWEnabled(mBlock.Header())

File: core/rootblockchain.go [hunk 17]
```diff
@@ -1423,15 +1415,15 @@ func (bc *RootBlockChain) GetRootChainStakesFunc() func(address account.Address,
...
+	if block.NumberU64() == 0 {
 		return nil, nil
 	}
-	if !bc.posw.IsPoSWEnabled(header) {
+	if !bc.posw.IsPoSWEnabled(block.Header()) {
```
Comment (pingke) [inline core/rootblockchain.go]: update bc.getSignedPoSWStakes(block.Header()) & bc.posw.IsPoSWEnabled(block.Header())

File: core/types/minorblock.go [hunk 18]
```diff
@@ -2,12 +2,13 @@
...
 
 import (
-	"github.com/QuarkChain/goquarkchain/params"
 	"math/big"
 	"sync/atomic"
 	"time"
 	"unsafe"
 
+	"github.com/QuarkChain/goquarkchain/params"
+
```
Comment (pingke) [inline core/types/minorblock.go]: remove emtpy line

File: core/types/rootblock.go [hunk 19]
```diff
@@ -291,7 +291,6 @@ func (b *RootBlock) TotalDifficulty() *big.Int      { return new(big.Int).Set(b.
 func (b *RootBlock) Nonce() uint64                  { return b.header.Nonce }
 func (b *RootBlock) Extra() []byte                  { return common.CopyBytes(b.header.Extra) }
 func (b *RootBlock) MixDigest() common.Hash         { return b.header.MixDigest }
-func (b *RootBlock) Signature() [65]byte            { return b.header.Signature }
```
Comment (pingke) [inline core/types/rootblock.go]: why remove?

Comment (516108736) [reply core/types/rootblock.go]: no used

**Pull request merged**